File: qasdk/src/main/java/co/astrnt/qasdk/utils/LogUtil.java
package co.astrnt.qasdk.utils;

import com.orhanobut.hawk.Hawk;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Locale;
import java.util.Set;
import java.util.TimeZone;

import co.astrnt.qasdk.constants.PreferenceKey;
import co.astrnt.qasdk.dao.LogDao;

public class LogUtil {

    public static void addNewLog(String interviewCode, LogDao itemLog) {
        if (interviewCode == null) return;
        List<LogDao> logDaoList = getLog(interviewCode);
        if (itemLog.getEvent().equals("Response API") || itemLog.getEvent().equals("Hit API")) {
            itemLog.setEvent(itemLog.getEvent() + " " + getLastApiCall());
        }
        logDaoList.add(itemLog);
        Set<LogDao> set = new LinkedHashSet<>(logDaoList);
        List<LogDao> logWithoutDuplicates = new ArrayList<>(set);
        Hawk.put(interviewCode, logWithoutDuplicates);
    }

    public static List<LogDao> getLog(String interviewCode) {
        if (interviewCode == null) {
            return new ArrayList<>();
        } else {
            List<LogDao> logDaoList = Hawk.get(interviewCode, new ArrayList<>());
            Set<LogDao> set = new LinkedHashSet<>(logDaoList);
            List<LogDao> logWithoutDuplicates = new ArrayList<>(set);
            Hawk.put(interviewCode, logWithoutDuplicates);
            return logWithoutDuplicates;
        }
    }

    public static String getTimeZone() {
        Calendar calendar = Calendar.getInstance(TimeZone.getTimeZone("GMT"), Locale.getDefault());
        String timeZone = new SimpleDateFormat("Z").format(calendar.getTime());
        return timeZone.substring(0, 3) + ":" + timeZone.substring(3, 5);
    }

    public static void clearSentLog(String interviewCode, List<LogDao> sentLog) {
        List<LogDao> logDaoList = getLog(interviewCode);
        logDaoList.removeAll(sentLog);
        Hawk.put(interviewCode, logDaoList);
    }

    public static String getLastApiCall() {
        return Hawk.get(PreferenceKey.KEY_LAST_API_CALL, null);
    }

    public static int getLastLogIndex() {
        return Hawk.get(PreferenceKey.KEY_LAST_LOG_INDEX, 0);
    }

    public static void saveLastLogIndex(int index) {
        Hawk.put(PreferenceKey.KEY_LAST_LOG_INDEX, index);
    }

}

File: qasdk/src/main/java/co/astrnt/qasdk/utils/LogUtil.kt
package co.astrnt.qasdk.utils

import co.astrnt.qasdk.constants.PreferenceKey
import co.astrnt.qasdk.dao.LogDao
import com.orhanobut.hawk.Hawk
import java.text.SimpleDateFormat
import java.util.*

object LogUtil {
    @JvmStatic
    fun addNewLog(interviewCode: String?, itemLog: LogDao) {
        if (interviewCode == null) return
        val logDaoList = getLog(interviewCode)
        if (itemLog.event == "Response API" || itemLog.event == "Hit API") {
            itemLog.event = itemLog.event + " " + lastApiCall
        }
        logDaoList.add(itemLog)
        val set: Set<LogDao> = LinkedHashSet(logDaoList)
        val logWithoutDuplicates: List<LogDao> = ArrayList(set)
        Hawk.put(interviewCode, logWithoutDuplicates)
    }

    @JvmStatic
    fun getLog(interviewCode: String?): MutableList<LogDao> {
        return if (interviewCode == null) {
            ArrayList()
        } else {
            val logDaoList: List<LogDao> = Hawk.get(interviewCode, ArrayList())
            val set: Set<LogDao> = LinkedHashSet(logDaoList)
            val logWithoutDuplicates: MutableList<LogDao> = ArrayList(set)
            Hawk.put<List<LogDao>>(interviewCode, logWithoutDuplicates)
            logWithoutDuplicates
        }
    }

    @JvmStatic
    val timeZone: String
        get() {
            val calendar = Calendar.getInstance(TimeZone.getTimeZone("GMT"), Locale.getDefault())
            val timeZone = SimpleDateFormat("Z").format(calendar.time)
            return timeZone.substring(0, 3) + ":" + timeZone.substring(3, 5)
        }

    @JvmStatic
    fun clearSentLog(interviewCode: String?, sentLog: List<LogDao>?) {
        val logDaoList = getLog(interviewCode)
        logDaoList.removeAll(sentLog!!)
        Hawk.put<List<LogDao>>(interviewCode, logDaoList)
    }

    val lastApiCall: String?
        get() = Hawk.get<String?>(PreferenceKey.KEY_LAST_API_CALL, null)
    @JvmStatic
    val lastLogIndex: Int
        get() = Hawk.get(PreferenceKey.KEY_LAST_LOG_INDEX, 0)

    @JvmStatic
    fun saveLastLogIndex(index: Int) {
        Hawk.put(PreferenceKey.KEY_LAST_LOG_INDEX, index)
    }
}